cli/connect/console.go

```diff
@@ -0,0 +1,352 @@
+	historyLines    []string
+	searchMode      bool
+
+	prefix            string
+	livePrefixEnabled bool
+	livePrefix        string
+	livePrefixFunc    func() (string, bool)
+
+	connOpts   *ConnOpts
+	conn       net.Conn
+	binaryConn *tarantool.Connection
+
+	executor  func(in string)
+	completer func(in prompt.Document) []prompt.Suggest
+
+	protocol   Protocol
+	outputMode ConsoleOutputMode
+
+	luaState *lua.LState
+
+	prompt *prompt.Prompt
+}
+
+func NewConsole(connOpts *ConnOpts, title string) (*Console, error) {
+	console := &Console{
+		title:      title,
+		outputMode: ConsoleYAMLOutput,
+		connOpts:   connOpts,
+		luaState:   lua.NewState(),
+	}
+
+	var err error
+
+	loadHistory(console)
+
+	console.conn, err = net.Dial(connOpts.Network, connOpts.Address)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to dial: %s", err)
+	}
+
+	greeting, err := readGreeting(console.conn)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read Tarantool greeting: %s", err)
+	}
+
+	switch {
+	case strings.Contains(greeting, "(Lua console)"):
+		console.protocol = PlainTextProtocol
+	case strings.Contains(greeting, "(Binary)"):
+		console.protocol = BinaryProtocol
+		if err := binaryConnect(console); err != nil {
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf("Unknown protocol: %s", greeting)
+	}
+
+	setTitle(console)
+	setPrefix(console)
+
+	console.executor, err = getExecutor(console)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get executor: %s", err)
+	}
+
+	console.completer, err = getCompleter(console)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get completer: %s", err)
+	}
+
+	options := getPromptOptions(console)
+
+	console.prompt = prompt.New(
+		console.executor,
+		console.completer,
+		options...,
+	)
+
+	return console, nil
+}
+
+func (console *Console) Run() error {
+	fmt.Printf("connected to %s\n", console.title)
+
+	pipedInputFound := false
+
+	pipedInputScanner := bufio.NewScanner(os.Stdin)
+	for pipedInputScanner.Scan() {
+		line := pipedInputScanner.Text()
+		console.executor(line)
+
+		pipedInputFound = true
+	}
+
+	if !pipedInputFound {
+		console.prompt.Run()
+	}
+
+	return nil
+}
+
+func (console *Console) Eval(funcBody string, args ...interface{}) (interface{}, error) {
+	var evalFunc func(console *Console, funcBodyFmt string, args ...interface{}) (interface{}, error)
+
+	switch {
+	case console.protocol == PlainTextProtocol:
+		evalFunc = plainTextEval
+	case console.protocol == BinaryProtocol:
+		evalFunc = binaryEval
+	default:
+		return nil, fmt.Errorf("Unknown protocol: %s", console.protocol)
+	}
+
+	return evalFunc(console, funcBody, args...)
+
+}
+
+func loadHistory(console *Console) {
+	var err error
+
+	homeDir, err := common.GetHomeDir()
+	if err != nil {
+		return
+	}
+
+	console.historyFilePath = filepath.Join(homeDir, HistoryFileName)
+
+	console.historyLines, err = common.GetLastNLines(console.historyFilePath, MaxHistoryLines)
+	if err != nil {
+		return
+	}
+}
```

Comment:
Maybe this function should return error or log it?

Reply:
The commands history is a very good feature, but it's not required to work with interactive console.
So, error shouldn't be returned.

But you are right, I will add debug log messages on console creation.